pkg/middleware/auth_proxy.go

```diff
@@ -41,32 +55,42 @@ func initContextWithAuthProxy(store *remotecache.RemoteCache, ctx *models.ReqCon
 		return true
 	}
 
-	// Try to log in user from various providers
-	id, err := auth.Login()
+	id, err := logUserIn(auth, username, logger, false)
 	if err != nil {
-		logger.Error(
-			"Failed to login",
-			"username", username,
-			"message", err.Error(),
-			"error", err.DetailsError,
-		)
 		ctx.Handle(407, err.Error(), err.DetailsError)
 		return true
 	}
 
-	// Get full user info
+	logger.Debug("Got user ID, getting full user info", "userID", id)
+
 	user, err := auth.GetSignedUser(id)
 	if err != nil {
-		logger.Error(
-			"Failed to get signed user",
-			"username", username,
-			"message", err.Error(),
-			"error", err.DetailsError,
-		)
-		ctx.Handle(407, err.Error(), err.DetailsError)
-		return true
+		// The reason we couldn't find the user corresponding to the ID might be that the ID was found from a stale
+		// cache entry. For example, if a user is deleted via the API, corresponding cache entries aren't invalidated
+		// because cache keys are computed from request header values and not just the user ID. Meaning that
+		// we can't easily derive cache keys to invalidate when deleting a user. To work around this, we try to
+		// log the user in again without the cache.
+		logger.Debug("Failed to get user info given ID, retrying without cache", "userID", id)
+		if err := auth.RemoveUserFromCache(logger); err != nil {
+			if !errors.Is(err, remotecache.ErrCacheItemNotFound) {
+				logger.Error("Got unexpected error when removing user from auth cache", "error", err)
+			}
+		}
+		id, err := logUserIn(auth, username, logger, true)
```

Comment:
```suggestion
		id, err = logUserIn(auth, username, logger, true)
```

need to override the outside variable `id` so as it will be set correctly [here](https://github.com/grafana/grafana/blob/fd49010775d8576b607e10bca52bd33b528c403e/pkg/middleware/auth_proxy.go#L99).

Reply:
Great catch @papagian, it's such an easy mistake in Go.